Optimisations

colinleroy opened the pull request:
Hello,

I hope you'll consider this PR. It aims at shaving a bit of performance (about 20% in redis-cache on my homepage) with two modifications:
- Avoid walking full arrays to determine group type: We use an associative array to get the type of the group instead of walking an array.
- Avoid sanitizing key/group multiple times in the same codepath: for example, ->get() was using sanitize_key to build the key and to check if the group was global.

I'll attach screenshots of KCachegrind that demonstrate the changes.

Hope this helps,
Colin

colinleroy commented:
Here is the call graph on WP_Object_Cache->get() before:
![image](https://user-images.githubusercontent.com/1016317/154808975-db498ce2-121b-4668-87cc-dc989f5bd650.png)

Here it is after:
![image](https://user-images.githubusercontent.com/1016317/154808988-c52a4220-a0be-43db-acd7-bb2701525891.png)

colinleroy commented:
Here is the total time passed in WP_Object_Cache functions before:
![image](https://user-images.githubusercontent.com/1016317/154809005-23fdd787-c300-4cc9-a79e-0c944d972073.png)

Here it is after:
![image](https://user-images.githubusercontent.com/1016317/154809018-f4e5c74e-ddf5-499b-b680-9e811d06ef0b.png)

You can see an improvement of about 19%.

tillkruss commented on includes/object-cache.php:
```diff
@@ -1175,10 +1198,14 @@ protected function add_or_replace( $add, $key, $value, $group = 'default', $expi
 
         // Save if group not excluded and redis is up.
-        if ( ! $this->is_ignored_group( $group ) && $this->redis_status() ) {
+        if ( ! $this->is_group_of_type( $san_group, "ignored" ) && $this->redis_status() ) {
```
For readability, let's keep `is_ignored_group()` and  `is_global_group()`.

colinleroy commented:
Hi @tillkruss ,
Thank you for your feedback. I've addressed your comment.
I also added a part in diagnostics to check on that array, and added refreshing the group type cache in case someone calls add_..._groups().

colinleroy commented:
Hi @tillkruss! Do you need me to fix something else on this PR?
Thanks :)
Colin

tillkruss commented:
@colinleroy: I have to review the changes in detail to make sure nothing breaks, otherwise 100,000 sites break.

colinleroy commented:
Ah, yes, understandably :)
Thanks !

tillkruss commented:
@colinleroy: Can you drop me an email via the chat bubble on https://objectcache.pro?

colinleroy commented:
Hi @tillkruss,

I'm sorry I've not been able to really get in touch with you yet... Life's a bit hectic. 
Do you still want to apply this PR ?

Thanks!
Colin

tillkruss commented:
@colinleroy: Yeah, this looks good.
